homework/Truong_Duong/Lesson_9/Javascript_4.md

```diff
@@ -0,0 +1,270 @@
+
+Use the === operator, or something equivalent such as indexOf, to determine whether two values are the same.
+Give the class a static from method that takes an iterable object as argument and creates a group that contains all the values produced by iterating over it.
+
+```js
+class Group {
+  _map = {};
+
+  static from(arrays) {
+    const group = new Group();
+
+    for (const ele of arrays) {
+      if (!(ele in group._map)) {
+        group._map[ele] = 1;
+      }
+    }
+    return group;
+  }
+
+  has(val) {
+    return val in this._map;
+  }
+
+  add(val) {
+    if (val in this._map) return;
+    this._map[val] = 1;
+  }
+
+  delete(val) {
+    if (!(val in this._map)) return;
+    delete this._map[val];
+  }
+
+}
+
+let group = Group.from([10, 20]);
+console.log(group.has(10)); // → true
+console.log(group.has(30)); // → false
+group.add(10);
+group.delete(10);
+console.log(group.has(10)); // → false
+```
+
+**Question 2**: Write an expression that matches only JavaScript-style numbers. It must support an optional minus or plus sign in front of the number, the decimal dot, and exponent notation—5e-3 or 1E10—again with an optional sign in front of the exponent. Also note that it is not necessary for there to be digits in front of or after the dot, but the number cannot be a dot alone. That is, .5 and 5. are valid JavaScript numbers, but alone dot isn’t.
+
+```js
+// Fill in this regular expression.
+let number = /^[+-]?((\d+[.]?\d*)|(\d*[.]?\d+))(e|e-|e\+|E|E-|E\+)?\d*$/;
+ 
+// Tests:
+for (let str of ["1", "-1", "+15", "1.55", ".5", "5.",
+                 "1.3e2", "1E-4", "1e+12"]) {
+  if (!number.test(str)) {
+    console.log(`Failed to match '${str}'`);
+  }
+}
+for (let str of ["1a", "+-1", "1.2.3", "1+1", "1e4.5",
+                 ".5.", "1f5", "."]) {
+  if (number.test(str)) {
+    console.log(`Incorrectly accepted '${str}'`);
+  }
+}
+```
+
+**Question 3**: Given an array of promises, `Promise.all` returns a promise that waits for all of the promises in the array to finish. It then succeeds, yielding an array of result values. **If a promise in the array fails, the promise returned by all fails too, with the failure reason from the failing promise.**
+Implement something like this yourself as a regular function called Promise_all.
+
+Remember that after a promise has succeeded or failed, it can’t succeed or fail again, and further calls to the functions that resolve it are ignored. This can simplify the way you handle failure of your promise.
+
+```js
+function Promise_all(promises) {
+  const results = Array(promises.length);
+  let resolveCount = 0;
+
+  return new Promise((resolve, reject) => {
+    if (promises.length === 0) resolve([]);
+
+    promises.map((promise, index) => {
```

Comment:
I think you should use `forEach` instead of `map` because `map` function will create and return new array

Reply:
Thanks for your advice 👍